### DISCUSSION TITLE
Struggling to create a suitable watertight mesh using Open3D from point cloud data

### DISCUSSION THREAD

**Discussion by conorssmith** [isl-org/Open3D#3913] [Q&A]
Hi there, I am trying to create a watertight mesh out of point cloud representing organ contour data from cone beam CT images. My goal is to take two meshes and calculate the volume of intersection between the two of them.

I have tried using each of the methods shown [here](http://www.open3d.org/docs/latest/tutorial/Advanced/surface_reconstruction.html)

###  Poisson Reconstruction

```
point_cloud = np.genfromtxt('ct_prostate_contour_data.csv', delimiter=',')
pcd = o3d.geometry.PointCloud()
pcd.points = o3d.utility.Vector3dVector(point_cloud)
pcd.compute_convex_hull()
pcd.estimate_normals()
pcd.orient_normals_consistent_tangent_plane(10)

mesh = o3d.geometry.TriangleMesh.create_from_point_cloud_poisson(pcd, depth=10, width=0, scale=20, linear_fit=True)[0]
mesh.compute_vertex_normals()
mesh.paint_uniform_color([0.5, 0.5, 0.5])
mesh.remove_degenerate_triangles()
o3d.visualization.draw_geometries([pcd, mesh], mesh_show_back_face=True)
```

While this method seemingly leads to a watertight mesh to my eye, the result of mesh.is_watertight() is False, however for the Bladder data it returns True. Furthermore, the algorithm extends the mesh above and below the vertical limits of the data. Wile this isn't a deal breaking issue if there were a way to minimize it that would be great. 

![Poisson Mesh](https://user-images.githubusercontent.com/88810617/129304982-e81b230a-ea7f-43be-ba71-df7418e2df8c.JPG)

### Ball Pivoting 

```
point_cloud = np.genfromtxt('ct_prostate_contour_data.csv', delimiter=',')
pcd = o3d.geometry.PointCloud()
pcd.points = o3d.utility.Vector3dVector(point_cloud)
pcd.compute_convex_hull()
pcd.estimate_normals()
pcd.orient_normals_consistent_tangent_plane(30)

distances = pcd.compute_nearest_neighbor_distance()
avg_dist = np.mean(distances)
radii = [0.1*avg_dist, 0.5*avg_dist, 1*avg_dist, 2*avg_dist] 
r = o3d.utility.DoubleVector(radii)
rec_mesh = o3d.geometry.TriangleMesh.create_from_point_cloud_ball_pivoting(pcd, r)
o3d.visualization.draw_geometries([pcd, rec_mesh], mesh_show_back_face=True)
```
This would be my preferred method if I were able to fill the holes as it simply connects vertices without interpolation.

![Ball Pivoting](https://user-images.githubusercontent.com/88810617/129305723-240adc9b-c43b-4f02-bfaf-dc22cd1dc373.JPG)

### Alpha Shapes
```
point_cloud = np.genfromtxt('ct_prostate_contour_data.csv', delimiter=',')
pcd = o3d.geometry.PointCloud()
pcd.points = o3d.utility.Vector3dVector(point_cloud)
alpha = 8
tetra_mesh, pt_map = o3d.geometry.TetraMesh.create_from_point_cloud(pcd)
mesh = o3d.geometry.TriangleMesh.create_from_point_cloud_alpha_shape(pcd, alpha, tetra_mesh, pt_map)
mesh.compute_vertex_normals()
mesh.paint_uniform_color([0.5, 0.5, 0.5])
mesh.remove_degenerate_triangles()
o3d.visualization.draw_geometries([pcd, mesh])
```

The results from this are similar to ball pivoting but worse. 

![Alpha Shapes](https://user-images.githubusercontent.com/88810617/129306147-9797b45a-4295-4dbb-8f43-87cab2b55a6b.JPG)

[ct_prostate_contour_data.csv](https://github.com/isl-org/Open3D/files/6979880/ct_prostate_contour_data.csv)
[ct_rectum_contour_data.csv](https://github.com/isl-org/Open3D/files/6979881/ct_rectum_contour_data.csv)
[ct_bladder_contour_data.csv](https://github.com/isl-org/Open3D/files/6979882/ct_bladder_contour_data.csv)

**Comment by Z-XQ** [comment]
same problem!

input mesh data: 

![image](https://github.com/isl-org/Open3D/assets/37925249/98f0abe5-21d8-4eb2-bc33-1af6a0de4a34)

full code:
            mesh_t = o3d.t.geometry.TriangleMesh.from_legacy(mesh)  # to tensor
            assert not mesh_t.to_legacy().is_watertight()  # true
            filled = mesh_t.fill_holes()
            assert filled.to_legacy().is_watertight()  # false

the mesh is still not watertight after using fill_holes().

**Comment by leolei1113** [comment]
Facing the same problem. Can anyone help? All of these 3 surface reconstriction methods fail to present decent results.

**Comment by germanros1987** [comment]
We are working on this. We will provide a solution asap.

**Comment by benjaminum** [comment]
Hi,
`create_from_point_cloud_poisson()` is the best choice for creating watertight meshes. 
Unfortunately there is no guarantee that the extracted mesh after optimization is watertight.
Something that can help is to reduce the resolution. Setting `depth=9` generates watertight meshes for the 3 point clouds.

**Comment by porpita1342** [comment]
@germanros1987 Hello. Are there any updates on this?

**Comment by lee-deraud** [comment]
Not sure that's possible in open3d alone. I'd try something like this:
1. convert pointclouds to mesh, whatever method looks best, and export
2. import and repair in Meshmixer or Fusion, again whatever works best. I've hd good luck with Fusion, but extreme cases, requiringthe 'Wrap' method, can take ages, i.e. minutes/hours rather than seconds.
3. import to OpenSCAD using one of the newer "nightly" builds, NOT the "last stable", which is from years ago). Take the intersection and export it.
4. That may or may not be watertight, but anothe pass through step 2 should handle that. (It'll be much faster this time.)
5. import back into open2d to get volume/surface area, whatever.
